ourprojects/models.py

```diff
@@ -30,20 +72,95 @@ class Meta(Record.Meta, TracksRun.Meta, TracksUpdates.Meta):
+    references: Reference = models.ManyToManyField(
+        "Reference", related_name="project_references"
+    )
+    """References associated with this project."""
     artifacts: Artifact = models.ManyToManyField(
         Artifact, through="ArtifactProject", related_name="Projects"
     )
     """Artifacts labeled with this Project."""
 
 
+class Reference(Record, CanCurate, TracksRun, TracksUpdates, ValidateFields):
+    """References such as a publication or document, with unique identifiers and metadata.
+
+    Example:
+        >>> reference = Reference(
+        ...     name="A Paper Title",
+        ...     abbr="APT",
+        ...     url="https://doi.org/10.1000/xyz123",
+        ...     pubmed_id=12345678,
+        ...     doi="10.1000/xyz123",
+        ...     preprint=False,
+        ...     journal="Nature Biotechnology",
+        ...     description="A groundbreaking research paper.",
+        ...     text="A really informative abstract.",
+        ...     published_at=date(2023, 11, 21),
+        ... ).save()
+    """
+
+    class Meta(Record.Meta, TracksRun.Meta, TracksUpdates.Meta):
+        abstract = False
+
+    id: int = models.AutoField(primary_key=True)
+    """Internal id, valid only in one DB instance."""
+    uid: str = CharField(
+        unique=True, max_length=12, db_index=True, default=ids.base62_12
+    )
+    """Universal id, valid across DB instances."""
+    name: str = CharField(db_index=True)
+    """Title or name of the reference document."""
+    abbr: str | None = CharField(
+        max_length=32,
+        db_index=True,
+        unique=True,
+        null=True,
+    )
+    """A unique abbreviation for the reference."""
+    url: str | None = URLField(null=True)
+    """URL linking to the reference."""
+    pubmed_id: int | None = BigIntegerField(null=True, db_index=True)
+    """A PudMmed ID."""
+    doi: str | None = CharField(
+        null=True,
+        db_index=True,
+        validators=[
+            RegexValidator(
+                regex=r"^(?:https?://(?:dx\.)?doi\.org/|doi:|DOI:)?10\.\d+/.*$",
+                message="Must be a DOI (e.g., 10.1000/xyz123 or https://doi.org/10.1000/xyz123)",
+            )
+        ],
+    )
+    """Digital Object Identifier (DOI) for the reference."""
+    preprint: bool = BooleanField(default=False)
+    """Whether the reference is from a preprint."""
+    journal: str | None = TextField(null=True)
+    """Name of the journal."""
+    description: str | None = TextField(null=True)
+    """Description of the reference."""
+    text: str | None = TextField(null=True)
+    """Abstract or full text of the reference."""
+    published_at: date | None = DateField(null=True, default=None)
+    """Publication date."""
+    people: Person = models.ManyToManyField(Person, related_name="references")
+    """All people associated with this reference."""
+    artifacts: Artifact = models.ManyToManyField(Artifact, related_name="references")
```

Comment:
Here you need a link model via through=ArtifactReference, follow the ArtifactProject example.

This is to enable labeling stratified by features.

Reply:
I thought that's what I had and just removed? :thinking: